File: training-data/ruby/fractional_knapsack.rb
Item = Struct.new(:name, :value, :weight) do
  def ratio = value.fdiv(weight)
end

def fractional_knapsack(items, capacity)
  items.sort_by { |i| -i.ratio }.each_with_object({ total: 0.0, taken: {} }) do |item, acc|
    break acc if capacity <= 0

    take = [item.weight, capacity].min
    acc[:total] += item.value * take.fdiv(item.weight)
    acc[:taken][item.name] = take.fdiv(item.weight).round(3)
    capacity -= take
  end
end

p fractional_knapsack([Item.new('gold', 60, 10), Item.new('silver', 100, 20), Item.new('bronze', 120, 30)], 50)

File: training-data/rust/fractional_knapsack.rs
struct Item {
    name: &'static str,
    value: f64,
    weight: f64,
}

fn fractional_knapsack(items: &mut [Item], mut capacity: f64) -> (f64, Vec<(&'static str, f64)>) {
    items.sort_by(|a, b| (b.value / b.weight).total_cmp(&(a.value / a.weight)));
    let mut total = 0.0;
    let mut taken = Vec::new();
    for it in items.iter() {
        if capacity <= 0.0 {
            break;
        }
        let take = it.weight.min(capacity);
        total += it.value * take / it.weight;
        capacity -= take;
        taken.push((it.name, take / it.weight));
    }
    (total, taken)
}

fn main() {
    let mut items = [
        Item { name: "gold", value: 60.0, weight: 10.0 },
        Item { name: "silver", value: 100.0, weight: 20.0 },
        Item { name: "bronze", value: 120.0, weight: 30.0 },
    ];
    let (total, taken) = fractional_knapsack(&mut items, 50.0);
    println!("total {}", total);
    for (name, frac) in taken {
        println!("{}: {:.0}%", name, frac * 100.0);
    }
}
